### PR TITLE
[Android] Adds `namespace` to module build file templates

### LINKED ISSUE #126403: [add-to-app] error : Namespace not specified. Please specify a namespace in the module's build.gradle
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Create a new android app and add a new flutter module to it using steps in option 2 : https://docs.flutter.dev/add-to-app/android/project-setup

run 'flutter pub get' in module to build .android folder

sync project in android studio

### Expected results

build is successful

### Actual results

app fails to build.  

Android studio is looking for 'namespace' setting in .android/build.gradle. adding a value manually makes the app build but this is a generated .android folder so any further runs of 'flutter pub get' to update dependencies remove any edits.

Error happens with 3.7.12-stable and 3.10.0-1.5.pre-beta

error output:

``` Script '/Users/xxx/.asdf/installs/flutter/3.12.7-stable/packages/flutter_tools/gradle/module_plugin_loader.gradle' line: 59

* What went wrong:
A problem occurred configuring root project 'Android'.
> A problem occurred configuring project ':flutter'.
   > Could not create an instance of type com.android.build.api.variant.impl.LibraryVariantBuilderImpl.
      > Namespace not specified. Please specify a namespace in the module's build.gradle file like so:
        
        android {
            namespace 'com.example.namespace'
        }
        
        If the package attribute is specified in the source AndroidManifest.xml, it can be migrated automatically to the namespace value in the build.gradle file using the AGP Upgrade Assistant; please refer to https://developer.android.com/studio/build/agp-upgrade-assistant for more information.

* Try:
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.7.12, on macOS 13.2 22D49 darwin-arm64, locale en-GB)
    • Flutter version 3.7.12 on channel stable at /Users/xxxx/.asdf/installs/flutter/3.7.12-stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4d9e56e694 (3 weeks ago), 2023-04-17 21:47:46 -0400
    • Engine revision 1a65d409c7
    • Dart version 2.19.6
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/xxxx/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.78.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (3 available)
    • iPhone 14 Pro (mobile) • D4CEE1AF-4892-48B9-A9D2-80C1D5F6175A • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 13.2 22D49 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 112.0.5615.137

[✓] HTTP Host Availability
    • All required HTTP hosts are available```

</details>

### PR DESCRIPTION
**Pull request by camsim99**
Adds `namespace` to module `build.gradle` file templates.

Fixes https://github.com/flutter/flutter/issues/126403.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by camsim99** [comment]
Looks like some tests are failing because they need multidex support. Not sure how they are related but I can update this PR

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart [hunk 1]
```diff
@@ -2919,6 +2919,45 @@ void main() {
...
+        moduleFlutterBuildGradleFilePath,
+      ],
+    );
+
+    final String moduleBuildGradleFileContent = await globals.fs.file(globals.fs.path.join(projectDir.path, moduleBuildGradleFilePath)).readAsString();
+    final String moduleAppBuildGradleFileContent = await globals.fs.file(globals.fs.path.join(projectDir.path, moduleAppBuildGradleFlePath)).readAsString();
+    final String moduleFlutterBuildGradleFileContent = await globals.fs.file(globals.fs.path.join(projectDir.path, moduleFlutterBuildGradleFilePath)).readAsString();
+
+    // Each build file should contain namespace.
+    expect(moduleBuildGradleFileContent.contains("namespace 'com.bar.foo.flutter_project'"), true);
```
Comment (reidbaker) [inline packages/flutter_tools/test/commands.shard/permeable/create_test.dart:2940]: Consider making this a constant.

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart [hunk 2]
```diff
@@ -2919,6 +2919,45 @@ void main() {
...
+    expect(moduleAppBuildGradleFileContent.contains("namespace 'com.bar.foo.flutter_project'"), true);
+    expect(moduleFlutterBuildGradleFileContent.contains("namespace 'com.bar.foo.flutter_project'"), true);
+
+    // The namespaces should be conditionalized for AGP <4.2.
+    expect(moduleBuildGradleFileContent.contains('if (project.android.hasProperty("namespace")) {'), true);
```
Comment (reidbaker) [inline packages/flutter_tools/test/commands.shard/permeable/create_test.dart:2945]: same here.

**Review by reidbaker — APPROVED** [review]

**Review by reidbaker — APPROVED** [review]

**Review by gmackall — APPROVED** [review]

**Comment by stuartmorgan** [comment]
> It is not some plugin

It is a module, which means it's intended for use in projects we don't directly control, so has the same fundamental compatibility concerns.

> Android devs tend to keep their projects up to date better than Flutter ones and definitely won't use such old versions of AGP as pre-4.2.

Do you have metrics showing that nobody using the module build has a project with AGP 4.1?

**Comment by vbuberen** [comment]
> It is a module, which means it's intended for use in projects we don't directly control, so has the same fundamental compatibility concerns.

We already had this discussion with you in Plus Plugins, so don't want to continue and would prefer to agree that we disagree here on this topic.

> Do you have metrics showing that nobody using the module build has a project with AGP 4.1?

I didn't say nobody. No, I don't have such metric, but it is something that you can get from the team responsible for releases of AGP as they might have such info. But as Android developer I know that sticking to old AGP in your project means that you consciously not using the latest optimisations for builds and new features. Also, `build.gradle` of your Android project is the same things as `pubspec.yaml` for Flutter devs and Android devs update dependencies there, including AGP (which is also updated with every release of Android Studio). In any of Android native projects I consulted or worked on I have seen `build.gradle` having AGP version older than 1 year. 4.2 is already 2 years old, thus, people with < 4.2 using something really old and it is hard to call a maintained project.

**Comment by reidbaker** [comment]
> > It is a module, which means it's intended for use in projects we don't directly control, so has the same fundamental compatibility concerns.
> 
> We already had this discussion with you in Plus Plugins, so don't want to continue and would prefer to agree that we disagree here on this topic.
> 
> > Do you have metrics showing that nobody using the module build has a project with AGP 4.1?
> 
> I didn't say nobody. No, I don't have such metric, but it is something that you can get from the team responsible for releases of AGP as they might have such info. But as Android developer I know that sticking to old AGP in your project means that you consciously not using the latest optimisations for builds and new features. Also, `build.gradle` of your Android project is the same things as `pubspec.yaml` for Flutter devs and Android devs update dependencies there, including AGP (which is also updated with every release of Android Studio). In any of Android native projects I consulted or worked on I have seen `build.gradle` having AGP version older than 1 year. 4.2 is already 2 years old, thus, people with < 4.2 using something really old and it is hard to call a maintained project.

Thank you for your engagement and concern for flutter. In this case there are 2 reasons why I encouraged @camsim99 to include agp 4.1 backwards compatibility. First is that we know about the compatibility issue and to keep compatibility is safe, easy to maintain, and straightforward to maintain. That alone would be a good reason for writing compatible code. The second is that flutter support many kinds of developers. Educational use including students and teachers, side projects,  ios/mac/windows only developers that want to run on android and developers that maintain dozens or hundreds of apps which may or may not ever be published by any app store. As a philosophy we try to make it easy for all kinds of developers to adopt new versions of flutter, new versions of packages and build new apps with the least amount of manual maintenance as we can get away with. 

So for this cl compatibility is in the spirit/philosophy of the flutter project.

**Pull request merged**